engine/runtime/src/main/java/org/enso/interpreter/runtime/scope/ModuleScope.java

```diff
@@ -262,4 +295,43 @@ public void reset() {
     conversions = new HashMap<>();
     polyglotSymbols = new HashMap<>();
   }
+
+  /**
+   * Create a copy of this `ModuleScope` while taking into account only the provided list of types.
+   *
+   * @param typeNames list of types to copy to the new scope
+   * @return a copy of this scope modulo the requested types
+   */
+  public ModuleScope withTypes(Collection<String> typeNames) {
+    Map<String, Object> polyglotSymbols = new HashMap<>(this.polyglotSymbols);
+    Map<String, Type> types = new HashMap<>(this.types);
+    Map<Type, Map<String, Function>> methods = new HashMap<>();
+    Map<Type, Map<Type, Function>> conversions = new HashMap<>();
+    Set<ModuleScope> imports = new HashSet<>(this.imports);
+    Set<ModuleScope> exports = new HashSet<>(this.exports);
+    this.types
+        .entrySet()
+        .forEach(
+            entry -> {
+              if (typeNames.contains(entry.getKey())) {
+                types.put(entry.getKey(), entry.getValue());
```

Comment:
`this.types` vs. `types` - that's quite cryptic style to overload names of instance and local variables. I don't think it contributes to easy readability.

I'd also much rather use `for (var entry : this.types.entrySet())` than these lambda callbacks (as they are harder to debug), but I know that's just my preference.